michbl/assignment2/BinaryTree.hpp

```diff
@@ -0,0 +1,194 @@
+                success = true;
+                return;
+            }
+            else if (node->right == nullptr) {
+                Node<T> *child = new Node<T>;
+                child->key = key;
+                node->right = child;
+                success = true;
+                return;
+            }
+            else { // parent already has 2 children
+                return;
+            }
+        }
+        insert(node->left, key, parentKey, success);
+        if (success) {
+            return;
+        }
+        insert(node->right, key, parentKey, success);
+    }
+
+    void destroyTree(Node<T> *node) {
+        if (node == nullptr) {
+            return;
+        }
+        destroyTree(node->left);
+        destroyTree(node->right);
+        delete node;
+    }
+
+    void printTree(Node<T> *node) {
+        if (node == nullptr) {
+            return;
+        }
+        printTree(node->left);
+        printTree(node->right);
+        cout << node->key << " ";
+    }
+
+    void printAncestors(Node<T> *node, T key, bool &found) {
+        if (node == nullptr) {
+            return;
+        }
+        if (node->key == key) {
+            found = true;
+            return;
+        }
+        if (!found) {
+            printAncestors(node->left, key, found);
+        }
+        if (!found) {
+            printAncestors(node->right, key, found);
+        }
+        if (found) {
+            if (node == this->root) {
+                cout << node->key << endl;
+            }
+            else {
+                cout << node->key << ", ";
+            }
+        }
+    }
+
+    bool bothKeysAreInSubtree(Node<T> *node, T key1, T key2) {
+        bool found1 = false, found2 = false;
+        bothKeysAreInSubtreeAux(node->left, key1, key2, found1, found2);
+        bothKeysAreInSubtreeAux(node->right, key1, key2, found1, found2);
+        return found1 && found2;
+    }
+
+    void bothKeysAreInSubtreeAux(Node<T> *node, T key1, T key2, bool &found1, bool &found2) {
+        if (node == nullptr) {
+            return;
+        }
+        if (node->key == key1) {
+            found1 = true;
+        }
+        if (node->key == key2) {
+            found2 = true;
+        }
+        bothKeysAreInSubtreeAux(node->left, key1, key2, found1, found2);
+        bothKeysAreInSubtreeAux(node->right, key1, key2, found1, found2);
+    }
+
+    void printLowestCommonAncestor(Node<T> *node, T key1, T key2, bool &found) {
+        if (node == nullptr) {
+            return;
+        }
+        if (!found) {
+            printLowestCommonAncestor(node->left, key1, key2, found);
+        }
+        if (!found) {
+            printLowestCommonAncestor(node->right, key1, key2, found);
+        }
+        if (!found && bothKeysAreInSubtree(node, key1, key2)) {
+            found = true;
+            cout << node->key << endl;
+        }
+    }
+
+
+public:
+    BinaryTree() {
+        this->root = nullptr;
+    }
+    BinaryTree(Node<T> *root) {

```

Comment:
At this moment, nodes are encapsulated in the tree and nothing returns them, yet this function gets a node as parameter. I don't think this constructor can ever be used at its current signature (unless constructing nodes outside of a tree, but I don't see a reason why to encourage the users to do so.

Is that a copy constructor?
If so, then its signature should be `BinaryTree(const Node<T>& root)`.
Anyway, this is optional because this assignment does not require you to copy trees :)